tests/saw/spec/handshake/s2n_handshake_io.cry

```diff
@@ -465,9 +711,20 @@ valid_connection conn = valid_handshake (conn.handshake)
 
 // Tells if the connection struct is in a valid initial state
 initial_connection : connection -> Bit
-initial_connection conn = conn.handshake.handshake_type == INITIAL /\ conn.handshake.message_number == 0
+initial_connection conn =
+       conn.handshake.handshake_type < `(S2N_HANDSHAKES_COUNT)
+    /\ conn.handshake.message_number == 0
     /\ (conn.corked == 0 \/ conn.corked == 1)
-    
+       // If using early data, a full handshake (i.e., chosen_psk being null)
+       // cannot happen
+    /\ ((conn.early_data_state == S2N_EARLY_DATA_ACCEPTED) ==> ~conn.chosen_psk_null)
+       // HelloRetryRequest and EarlyData cannot both happen simultaneously
+    /\ ~(((conn.handshake.handshake_type && HELLO_RETRY_REQUEST) != zero)
+            /\ (conn.early_data_state == S2N_EARLY_DATA_ACCEPTED))
+       // EarlyClientCCS implies MiddleboxCompat
```

Comment:
> Are these assumptions worth more documentation? Without context, it seems strange that a proof of the TLS1.3 handshakes makes assumptions about the TLS1.3 handshakes. We should probably document that it's because some of the logic has moved out of the functions these proofs verify.

Yes, I agree that some additional documentation on this point would be nice. I've pushed a commit attempting to do so.

> And since this does weaken the proof, maybe we should open an issue to try to move that logic back into the scope of the proof? Like maybe set_handshake_type should verify that the already set handshake type flags are valid.

Sure, I'd be fine with that. If I understand correctly, the [`valid_connection`](https://github.com/aws/s2n-tls/blob/abed2a324e59da13f73967ed7ea936b5dd2a1aae/tests/saw/spec/handshake/s2n_handshake_io.cry#L463-L464) function checks if the `handshake_type` flags are valid, so one could imagine replacing some of these assumptions with `valid_connection` instead. I tried to avoid using `valid_connection` since it's not obvious at a glance which handshakes it excludes, whereas stating the assumptions in terms of `handshake_type` flags seemed more intuitive. That being said, I don't have too strong an opinion on this, and I'd be happy with defining `initial_connection` in some alternative way.

Reply:
I made a proof for a potential mitigation: https://github.com/aws/s2n-tls/issues/3047

Reply:
Thanks! I've pushed a commit which mentions this issue in the comments.